File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target){
        int left = 0;
        int right = nums.size()-1;

        while(left <= right){
            int mid = left + (right - left)/2;

            if(target == nums[mid]){
                return mid;
            }else if(target < nums[mid]){
                right = mid - 1;
            }else if(target > nums[mid]){
                left = mid + 1;
            }
        }

        return left;
    }
    int lengthOfLIS(vector<int>& nums) {
        vector<int>ans;

        for(int i = 0; i < nums.size(); i++){
            if(ans.empty() || ans.back() < nums[i]){
                ans.push_back(nums[i]);
            }else{
                int idx = binarySearch(ans, nums[i]);
                ans[idx] = nums[i];
            }
        }

        return ans.size();
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def bs(self, nums:List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = left + (right - left)//2

            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        
        return left

    def lengthOfLIS(self, nums: List[int]) -> int:
        res = []

        for i in range(len(nums)):
            if len(res) == 0 or res[-1] < nums[i]:
                res.append(nums[i])
            else:
                idx = self.bs(res, nums[i])
                res[idx] = nums[i]

        return len(res)